doc/tutorial/scalajs-vite.md

```diff
@@ -0,0 +1,416 @@
+Is that also the case if we change the Scala source files?
+
+Indeed it is.
+Let us change the message to
+
+{% highlight diff %}
+       <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript" target="_blank">
+         <img src="/javascript.svg" class="logo vanilla" alt="JavaScript logo" />
+       </a>
+-      <h1>Hello Scala.js!</h1>
++      <h1>Hello Scala.js and Vite!</h1>
+       <div class="card">
+         <button id="counter" type="button"></button>
+       </div>
+{% endhighlight %}
+
+Once we save, we notice that the browser refreshes with the updated message.
+
+There are two things happening behind the scenes:
+
+1. The `~fastLinkJS` task in sbt notices that a `.scala` file has changed, and therefore rebuilds the `.js` output.
+1. The `npm run dev` process with Vite notices that a `.js` file imported from `/main.js` has changed, and triggers a refresh with the updated files.
+
+All these steps are *incremental*.
+When we change a single Scala file, only that one gets recompiled by the Scala incremental compiler.
+Then, only the affected small `.js` modules produced by `fastLinkJS` are regenerated.
+Finally, Vite only reloads those small `.js` files that were touched.
+This ensures that the development cycle remains as short as possible.
+
+## Production build
+
+The `fastLinkJS` task of sbt and the `npm run dev` task of Vite are optimized for incremental development.
+For production, we want to perform more optimizations on the Scala.js side with `fullLinkJS`, and bundle minimized files with `npm run build`.
+
+For that to work well, we need to conditionally wire the output of `fastLinkJS` (resp. `fullLinkJS`) to the input of Vite when it is in development mode (resp. production mode).
+This requires some configuration in Vite.
+
+We create the file `vite.config.js` at the root of our project, with the following content:
+
+{% highlight javascript %}
+import { spawnSync } from "child_process";
+import { defineConfig } from "vite";
+
+// Detect whether Vite runs in development or production mode
+function isDev() {
+  return process.env.NODE_ENV !== "production";
+}
+
+// Utility to invoke a given sbt task and fetch its output
+function printSbtTask(task) {
+  const args = ["--error", "--batch", `print ${task}`];
+  const options = {
+    stdio: [
+      "pipe", // StdIn
+      "pipe", // StdOut
+      "inherit", // StdErr
+    ],
+  };
+  const result = process.platform === 'win32'
+    ? spawnSync("sbt.bat", args.map(x => `"${x}"`), {shell: true, ...options})
+    : spawnSync("sbt", args, options);
+
+  if (result.error)
+    throw result.error;
+  if (result.status !== 0)
+    throw new Error(`sbt process failed with exit code ${result.status}`);
+  return result.stdout.toString('utf8').trim();
+}
+
+// Get the output of fastLinkJS or fullLinkJS depending on isDev()
+const scalaJSOutputTask = isDev() ? "fastLinkJSOutput" : "fullLinkJSOutput";
+const scalaJSOutput = printSbtTask(scalaJSOutputTask);
```

Comment:
> Have you considered a vite / rollup plugin (potentially in-place)? I realize that this is potentially significant scope creep, but I think for the first tutorial on the main page we should be better than that.

I have considered that. It was one of my first ideas. But then I internally got as critique that it would not help Mill users (or other build tools). As is, the Mill user can probably figure out that they can invoke Mill instead of sbt. If we make a vite-sbt plugin, this is hidden and not available to Mill users.

That said, you're not the first one also saying that this looks jarring as getting-started experience. So I'm taking it as a sign that it's probably worth doing the vite-sbt integration better, at the cost of being less friendly to other build tools.

Reply:
> That said, you're not the first one also saying that this looks jarring as getting-started experience. So I'm taking it as a sign that it's probably worth doing the vite-sbt integration better, at the cost of being less friendly to other build tools.

FWIW: If the plugin is reasonably simple, it might make sense to go over how it works in the tutorial. So it's easier to port it to mill.

Or even inline the plugin into the build (if it can go into a separate file, it's maybe not too bad and it'll make it much easier to, say, deal with multiple Scala.js outputs for say FE / BE).